### ISSUE TITLE
linking libcuda.so.1 in a Docker image at build time

### ISSUE THREAD

**Issue by masonk** [tensorflow/tensorflow#25865]
**System information**
- Host system Ubuntu 18.04
- TensorFlow r1.13 branch
- Installing from `nvidia/cuda:10.0-cudnn7-devel-ubuntu16.04` base docker image
- Bazel 0.21.0
- GCC/Compiler version (if compiling from source):
- Cuda 10.0, cuDNN 7.4
- RTX 2080 Ti

**Describe the problem**
When building `//tensorflow/libtensorflow.so` in a RUN step of my Dockerfile, I get this error:

```
ERROR: /tensorflow/tensorflow/cc/BUILD:477:1: Linking of rule '//tensorflow/cc:ops/data_flow_ops_gen_cc' failed (Exit 1)
/usr/bin/ld: warning: libcuda.so.1, needed by bazel-out/host/bin/_solib_local/_U_S_Stensorflow_Scc_Cops_Sdata_Uflow_Uops_Ugen_Ucc___Utensorflow/libtensorflow_framework.so, not found (try using -rpath or -rpath-link)
<snip>
Target //tensorflow:libtensorflow.so failed to build
ERROR: /tensorflow/tensorflow/cc/BUILD:477:1 Linking of rule '//tensorflow/cc:ops/nn_ops_gen_cc' failed (Exit 1): crosstool_wrapper_driver_is_not_gcc failed: error executing command
  (cd /root/.cache/bazel/_bazel_root/68a62076e91007a7908bc42a32e4cff9/execroot/org_tensorflow && \
  exec env - \
    PATH=/bin:/usr/bin \
    PWD=/proc/self/cwd \
  external/local_config_cuda/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc -o bazel-out/host/bin/tensorflow/cc/ops/nn_ops_gen_cc '-Wl,-rpath,$ORIGIN/../../../_solib_local/_U_S_Stensorflow_Scc_Cops_Snn_Uops_Ugen_Ucc___Utensorflow' '-Wl,-rpath,$ORIGIN/../../../_solib_local/_U@local_Uconfig_Ucuda_S_Scuda_Ccudart___Uexternal_Slocal_Uconfig_Ucuda_Scuda_Scuda_Slib' -Lbazel-out/host/bin/_solib_local/_U_S_Stensorflow_Scc_Cops_Snn_Uops_Ugen_Ucc___Utensorflow -Lbazel-out/host/bin/_solib_local/_U@local_Uconfig_Ucuda_S_Scuda_Ccudart___Uexternal_Slocal_Uconfig_Ucuda_Scuda_Scuda_Slib -Wl,-ldl '-Wl,-rpath,$ORIGIN/,-rpath,$ORIGIN/..,-rpath,$ORIGIN/../..' -pthread -Wl,-rpath,../local_config_cuda/cuda/lib64 -Wl,-rpath,../local_config_cuda/cuda/extras/CUPTI/lib64 -Wl,-S -Wl,-no-as-needed -pie -Wl,-z,relro,-z,now '-Wl,--build-id=md5' '-Wl,--hash-style=gnu' -no-canonical-prefixes -fno-canonical-system-headers -B/usr/bin -Wl,--gc-sections -Wl,@bazel-out/host/bin/tensorflow/cc/ops/nn_ops_gen_cc-2.params)
Execution platform: @bazel_tools//platforms:host_platform
INFO: Elapsed time: 229.527s, Critical Path: 139.75s
INFO: 2250 processes: 2250 local.
```

This only occurs when trying to build **in a RUN step**. If skip buliding and shell into the container, it builds to completion. There are several similar issues on github, two of the most similar are:

https://github.com/tensorflow/tensorflow/issues/16184
https://github.com/tensorflow/tensorflow/issues/14573
Particularly https://github.com/tensorflow/tensorflow/issues/14573#issuecomment-362424509

I have tried every variation of the advice given in that thread, as well as cribbed lines from this Dockerfile that is doing a similar thing:
https://github.com/gunan/tensorflow-docker/blob/master/gpu-devel/Dockerfile.ubuntu#L64

Yet, I'm still getting this message. 

I don't know what exactly nvidia is doing with its cuda libraries, but I think they are not in all the same places at buildtime and runtime. I discovered this discrepancy accidentally with `RUN find / | grep libcuda.so.1` in my Dockerfile:

```
/usr/local/cuda-10.0/compat/libcuda.so.1
```
This is the only place that libcuda.so.1 exists at build time. However, even if I symlink from `/usr/local/cuda-10.0/compat/libcuda.so.1
` to  a place that is definitely in `LD_LIBRARY_PATH`, _and_ make sure to pass `action_env=LD_LIBRARY_PATH=${LD_LIBRARY_PATH}`, I _still_ get this error. 

I'm completely stumped at this point.

To reproduce this you can use the dockerfile attached at the bottom of this post.

Build in the container - works:

```
docker build -f test.dockerfile -t masontest . && docker run --vm=nvidia --rm masontest 
cd /tensorflow
RUN bazel build \
 --verbose_failures \
 --spawn_strategy=standalone \
    --genrule_strategy=standalone \
    --action_env=LD_LIBRARY_PATH=${LD_LIBRARY_PATH} \
    //tensorflow:libtensorflow.so
```
Uncomment the bazel command - the image fails to build with the linker error above.

```
FROM nvidia/cuda:10.0-cudnn7-devel-ubuntu16.04
# TF docker (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/dockerfiles/dockerfiles/devel-gpu.Dockerfile)
# uses cuda:10.0-base-ubuntu16.04, but we need full sources 
# because later in this docker we are going to compile Tensorflow from scratch
# using bazel.

# so we aren't prompted to set up the keyboard and other nonsense
ENV DEBIAN_FRONTEND noninteractive 

RUN set -eux; \
    apt-get update && apt-get install -y --no-install-recommends \
    autoconf \
    automake \
    build-essential \
    ca-certificates \
    g++ \
    gcc \
    git \
    # these were needed by the mnistCUDNN demo
    libfreeimage3 \
    libfreeimage-dev \
    make \
    # bazel needs python2 https://github.com/tensorflow/tensorflow/issues/15618
    python \
    # tf needs swig
    swig \
    unzip \
    wget \
    zlib1g-dev; \
    apt-get clean; \
    rm -rf /var/lib/apt/lists/*

# Tensorflow pukes when you try this with 0.22.0
ENV BAZEL_VERSION 0.21.0
WORKDIR /
RUN mkdir /bazel && \
    cd /bazel && \
    wget https://github.com/bazelbuild/bazel/releases/download/$BAZEL_VERSION/bazel-$BAZEL_VERSION-installer-linux-x86_64.sh && \
    # curl -fSsL -o /bazel/LICENSE.txt https://raw.githubusercontent.com/bazelbuild/bazel/master/LICENSE.txt && \
    chmod +x bazel-*.sh && \
    ./bazel-$BAZEL_VERSION-installer-linux-x86_64.sh && \
    cd / && \
    rm -f /bazel/bazel-$BAZEL_VERSION-installer-linux-x86_64.sh

RUN echo $(g++ --version)
RUN echo $(bazel version)

### TENSORFLOW 
ENV TENSORFLOW_VERSION r1.13
RUN git clone https://github.com/tensorflow/tensorflow.git && \
    cd tensorflow && \
    git checkout ${TENSORFLOW_VERSION}

ENV PYTHON_LIB_PATH /usr/local/lib/python2.7/dist-packages
ENV PYTHONPATH /tensorflow/lib
ENV PYTHON_ARG /tensorflow/lib

ENV GCC_HOST_COMPILER_PATH /usr/bin/gcc
ENV CC_OPT_FLAGS="-march=native"
ENV CUDA_TOOLKIT_PATH /usr/local/cuda
ENV CUDNN_INSTALL_PATH /usr/lib/x86_64-linux-gnu
ENV TF_CUDA_COMPUTE_CAPABILITIES 7.0 
ENV TF_CUDNN_VERSION 7
ENV TF_CUDA_VERSION 10.0
ENV TF_NEED_GCP 0
ENV TF_ENABLE_XLA 0
ENV TF_NEED_OPENCL 0
ENV TF_NEED_CUDA 1
ENV TF_CUDA_VERSION 10.0.0
ENV TF_NEED_HDFS 0
ENV TF_NEED_TENSORRT 0
ENV TF_NCCL_VERSION 2
ENV TF_BUILD_BRANCH r1.13
ENV TF_NEED_ROCM 0
ENV TF_NEED_OPENCL_SYCL 0
ENV TMP /tmp/tensorflow
RUN echo "startup --batch" >>/etc/bazel.bazelrc

ENV LD_LIBRARY_PATH="/usr/local/lib:/usr/local/cuda/lib64:/usr/local/cuda/lib64/stubs:${LD_LIBRARY_PATH}"

RUN find / | grep  libcuda.so.1
WORKDIR /tensorflow
RUN yes '' | ./configure
# https://github.com/gunan/tensorflow-docker/blob/master/gpu-devel/Dockerfile.ubuntu#L64
# Is this ln -s is necessary when building? Doesn't seem to help, and the actual libcuda ... https://github.com/tensorflow/tensorflow/issues/14573#issuecomment-362424509
# Similarly, we need to workaround sandboxing issues: https://github.com/bazelbuild/bazel/issues/418
RUN ln -s /usr/local/cuda/lib64/stubs/libcuda.so /usr/local/cuda/lib64/stubs/libcuda.so.1
# RUN bazel build \
#  --verbose_failures \
#  --spawn_strategy=standalone \
#     --genrule_strategy=standalone \
#     --action_env=LD_LIBRARY_PATH=${LD_LIBRARY_PATH} \
#     //tensorflow:libtensorflow.so
```

**Comment by masonk** [comment]
Building this at head succeded, presumably because https://github.com/tensorflow/tensorflow/issues/16184 switched to dynamic loading of libcuda. It sure would be nice to have a working development Docker with the latest stable release but I understand that there are a lot of things that would be nice to have. I can't imagine that this is only broken for me - I'm just not doing anything special.

**Comment by masonk** [comment]
Related issue - original incidence of this issue and the originally proposed fix. Will try it tonight.

https://github.com/tensorflow/tensorflow/pull/13399

**Comment by masonk** [comment]
Also https://github.com/NVIDIA/nvidia-docker/issues/775

**Comment by angersson** [comment]
I've marked this as "Community Support" because the TF team doesn't officially support building inside a Docker build process.

@masonk, do the official [Docker images](https://hub.docker.com/r/tensorflow/tensorflow) not meet your use case? We provide images with the latest stable releases and with the nightly packages pre-installed with Nvidia GPU support.

**Comment by masonk** [comment]
@angersson 

I don't _think_ that the official images meet my use-case, because I am building a Rust app that uses the rust-tensorflow (bindings to the C API), and it needs to link against libtensorflow, which I did not find in the official images.

I was able to build GPU-enabled libtensorflow v1.12.0 from scratch using a customer Dockerfile, which I will reproduce here for posterity.  

```
FROM nvidia/cuda:9.0-cudnn7-devel-ubuntu16.04
# TF docker (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/dockerfiles/dockerfiles/devel-gpu.Dockerfile)
# uses cuda:10.0-base-ubuntu16.04, but we need full sources 
# because later in this docker we are going to compile Tensorflow from scratch
# using bazel.
# so we aren't prompted to set up the keyboard and other nonsense
ENV DEBIAN_FRONTEND noninteractive 

RUN set -eux; \
    apt-get update && apt-get install -y --no-install-recommends \
    autoconf \
    automake \
    build-essential \
    ca-certificates \
    g++ \
    gcc \
    git \
    less \
    # these were needed by the mnistCUDNN demo
    libfreeimage3 \
    libfreeimage-dev \
    # These two packages allow a TF to be built that run on multiple GPUs
    # libnccl2 libnccl-dev \
    # Seraphim depends on libssl via the openssl-sys crate (which I think is a transitive dep of ctrlc)
    libssl-dev \ 
    lldb-7 \
    gdbserver \
    make \
    # Rust libssl crate uses pkg-config to find the openssl system headers
    pkg-config \
    # bazel needs python2 https://github.com/tensorflow/tensorflow/issues/15618
    python \
    # tf needs swig
    swig \
    #just for my own convenience, I like this package
    tree \ 
    unzip \
    wget \
    zlib1g-dev; \
    apt-get clean; \
    rm -rf /var/lib/apt/lists/*

# This version has to correspond to the appropriate bazel version
# r1.12 can't be built with bazel 0.19 https://github.com/tensorflow/tensorflow/issues/23401#issuecomment-434681778
# It also can't be built with any higher version of bazel that I tried
# Note to self for the future: r1.13 can be built with bazel 0.21, but not 0.22
ENV BAZEL_VERSION 0.18.0
WORKDIR /
RUN mkdir /bazel && \
    cd /bazel && \
    wget https://github.com/bazelbuild/bazel/releases/download/$BAZEL_VERSION/bazel-$BAZEL_VERSION-installer-linux-x86_64.sh && \
    # curl -fSsL -o /bazel/LICENSE.txt https://raw.githubusercontent.com/bazelbuild/bazel/master/LICENSE.txt && \
    chmod +x bazel-*.sh && \
    ./bazel-$BAZEL_VERSION-installer-linux-x86_64.sh && \
    cd / && \
    rm -f /bazel/bazel-$BAZEL_VERSION-installer-linux-x86_64.sh

RUN echo $(g++ --version)
RUN echo $(bazel version)

### TENSORFLOW 
# rust-tensorflow works with r1.12
# https://github.com/tensorflow/tensorflow/issues/25865
ENV TENSORFLOW_VERSION r1.12
RUN git clone https://github.com/tensorflow/tensorflow.git && \
    cd tensorflow && \
    git checkout ${TENSORFLOW_VERSION}

ENV TF_NEED_CUDA 1

WORKDIR /tensorflow
RUN yes '' | ./configure
# https://github.com/gunan/tensorflow-docker/blob/master/gpu-devel/Dockerfile.ubuntu#L64
# This ln-s is necessary when building ... https://github.com/tensorflow/tensorflow/issues/14573#issuecomment-362424509
# Similarly, we need to workaround sandboxing issues:
#   https://github.com/bazelbuild/bazel/issues/418

# Context for this elaborate bazel invocation is here https://github.com/tensorflow/tensorflow/issues/25865
# TLDR: tf <=1.13 links to libcuda.so.1 at build time, but those headers aren't available in the nvidia docker
# at build time. (They are mounted in the running container, but not at build time). So we link to
# stubs to satisfy the build-time linker.
RUN export LD_LIBRARY_PATH=/usr/local/cuda/lib64/stubs:$LD_LIBRARY_PATH && \
    ln -s /usr/local/cuda/lib64/stubs/libcuda.so /usr/local/cuda/lib64/stubs/libcuda.so.1 && \
    bazel build \
    -c opt \
    --config="opt" \
    --copt=-mavx --copt=-mavx2 --copt=-mfma --copt=-mfpmath=both --copt=-msse4.2 \
    --action_env=LD_LIBRARY_PATH=${LD_LIBRARY_PATH} \
    --verbose_failures \
    --spawn_strategy=standalone \
    --genrule_strategy=standalone \
    //tensorflow:libtensorflow.so && \
    rm /usr/local/cuda/lib64/stubs/libcuda.so.1

RUN cp bazel-bin/tensorflow/libtensorflow.so /usr/local/lib


```
